Integrate sphinx starter pack

a-velasco opened the pull request:
This PR adds Canonical's Sphinx tooling to our documentation so that it can be rendered as a Read The Docs project, and edits the existing docs to remove build errors and polish the structure.

### Integration with the [Canonical Sphinx Starter Pack](https://github.com/canonical/sphinx-docs-starter-pack)

We add several files that enable our docs to be built with Sphinx, checked in our CI, and published on Read The Docs.

In the root directory:
* `.readthedocs.yaml`: Build configuration for the Read The Docs project

Of note in the `/docs` folder:
* `/.sphinx` folder: Build checks and tooling for links, spelling, inclusive language, and other utilities.
* `conf.py`: Sphinx configuration options, e.g. project name, associated links, features we want to enable/disable, HTML/CSS, MyST settings, extensions
* `Makefile`: Makefile for Sphinx's compiler

In the `.github/workflows` folder:
* `automatic-doc-checks.yml`: Reusable workflows from [canonical/documentation-workflows](https://github.com/canonical/documentation-workflows)
* `markdown-style-checks.yml`: Markdown linter

### Edits to content and structure

The existing files were restructured into Diataxis folders (for now, just `tutorial` and `how-to`), along with new index pages for each folder. The tutorial pages were united into a single page. 

The content of several pages was polished in miscellaneous ways, largely formatting-wise (i.e. headings, code blocks, and links/refs). Several titles and filenames were lightly edited for consistency with one another.

a-velasco committed:
Initial transfer of starter pack files

a-velasco committed:
Customize conf.py

a-velasco committed:
Organize markdown files and add index pages

a-velasco committed:
Clean up page titles

a-velasco committed:
Restructure tutorial into one page

a-velasco committed:
Polish formatting and content of monitoring guide

a-velasco committed:
Miscellaneous polishing

a-velasco committed:
Polish TLS docs

izmalk commented on .github/workflows/sphinx-python-dependency-build-checks.yml:
```diff
@@ -0,0 +1,51 @@
+# the packages required in the build environment to do that.
+#
+# This is needed because some projects embeds the documentation into built
+# artifacts which involves rendering the documentation on the target
```
Suggestion: Do you think we can remove this test completely? This particular check is probably redundant for this product, as it doesn't require us to build the Sphinx and docs on some weird edge architectures.

izmalk commented on docs/.custom_wordlist.txt:
```diff
@@ -0,0 +1,35 @@
+backend
+backends
+Charmcraft
+cjk
```
Question: Where do we need this?

a-velasco replied:
I am quite unsure about this file in general. I still don't know why we have this custom wordlist when we also have the Vale style definitions. Something to look into.

a-velasco committed:
Remove unnecessary workflow

skourta commented on docs/how-to/tls/index.md:
```diff
@@ -0,0 +1,15 @@
+:maxdepth: 3
+
+Enable TSL <enable-tls>
+Disable TSL <disable-tls>
```
TLS typo

skourta replied:
Is this one fixed?

skourta commented on docs/how-to/tls/disable-tls.md:
````diff
@@ -0,0 +1,100 @@
+
+To disable peer-to-peer communication, run:
+
+```bash
````
switch to {terminal} syntax for sphinx

a-velasco committed:
Ignore anchor link check for juju docs

a-velasco committed:
Fix syntax error

a-velasco committed:
Fix TLS folder navigation issue

a-velasco committed:
Remove unused paths

reneradoi reviewed (dismissed):
Thank you Andreia!

izmalk commented on .readthedocs.yaml:
```diff
@@ -0,0 +1,39 @@
+
+# If using Sphinx, optionally build your docs in additional formats such as PDF
+formats:
+- pdf
```
Suggestion: This can be commented out if you don't need PDFs.

izmalk commented on docs/index.md:
```diff
@@ -0,0 +1,25 @@
+# Charmed Etcd documentation
+
+Etcd is a distributed, reliable key-value store for the most critical data of distributed systems. This charmed operator deploys and operates etcd on virtual machines.
```
Suggestion: Can we add the first paragraph, explaining that Charmed Etcd is a Juju charm that automates, simplifies, etc. (pun not intended) ? It's important to clarify difference between the original product (Etcd) and our charmed operator.

a-velasco replied:
Adding this to list of content improvements to make in a new PR.

izmalk commented on docs/index.md:
```diff
@@ -0,0 +1,25 @@
+<!--
+| [Explanation](/explanation/index) </br> Concepts - discussion and clarification of key topics | [Reference](/reference/index) </br> Technical information - specifications, APIs, architecture |
+-->
+
```
Question: What about the Project and community section from the standard Home page template? I think it may be missing here.

a-velasco replied:
Will address in new PR.

izmalk commented on docs/index.md:
````diff
@@ -0,0 +1,25 @@
+
+```{toctree}
+:titlesonly:
+:maxdepth: 2
````
Question: Do you find this `maxdepth` option useful? I, personally, find it troublesome at times. I'm just curious on why you use it, maybe I'm missing out on something.

izmalk commented on docs/how-to/index.md:
```diff
@@ -0,0 +1,14 @@
+(how-to)=
+# How-to guides
+
```
Suggestion: This is an empty page. We should at least make the toctree not `hidden` or better yet - add some content to this as a landing page.

a-velasco replied:
Adding content to this page is planned for the next PR, along with some other toctree changes (including `maxdepth`). For some reason, disabling `hidden` here leads to weird behavior from the nav, so I will leave it as-is for now.

izmalk commented on docs/tutorial/index.md:
````diff
@@ -0,0 +1,190 @@
+{"etcdserver":"3.5.16","etcdcluster":"3.5.0"}
+```
+
+## Next steps
````
Suggestion: Add an empty line.

```suggestion
## Next steps

```

izmalk commented on docs/how-to/tls/index.md:
```diff
@@ -0,0 +1,14 @@
+(tls)=
+# TLS encryption
+
+Transport Layer Security (TLS) plays a crucial role in securing database communications.  Just as it protects web traffic, TLS encrypts the data transmitted between database clients and servers, preventing unauthorized access and ensuring confidentiality.
```
Suggestion: Add the protocol to define what TLS is and add a line break as best practice (rather than storing paragraphs of text in a single line, use line breaks between sentences to limit the scope of future changes to this content).

```suggestion
Transport Layer Security (TLS) protocol plays a crucial role in securing database communications. 
Just as it protects web traffic, TLS encrypts the data transmitted between database clients and servers, preventing unauthorized access and ensuring confidentiality.
```

izmalk commented on docs/how-to/enable-monitoring.md:
```diff
@@ -1,14 +1,21 @@
+(enable-monitoring)=
+# How to enable monitoring
+
 The Canonical Observability Stack (COS) is a set of tools that facilitates gathering, processing, visualizing, and setting up alerts on telemetry signals generated by workloads in and outside of Juju.
```
Suggestion: Add the link to the COS docs before the Grafana and Prometheus.

```suggestion
The [Canonical Observability Stack (COS)](https://charmhub.io/topics/canonical-observability-stack) is a set of tools that facilitates gathering, processing, visualizing, and setting up alerts on telemetry signals generated by workloads in and outside of Juju.
```

izmalk commented on docs/how-to/enable-monitoring.md:
```diff
@@ -1,14 +1,21 @@
-Deploy the [`cos-lite`](https://charmhub.io/topics/canonical-observability-stack) bundle in a Kubernetes environment following the [MicroK8s guide](https://charmhub.io/topics/canonical-observability-stack/tutorials/install-microk8s). Since the etcd charm is deployed directly on a cloud infrastructure environment, it requires offering the endpoints of the COS relations. The [offers-overlay](https: ...
 
-Once the COS bundle is deployed, you should have Grafana and Prometheus running in your Kubernetes cluster. Run `juju status` to see the status of the COS model. The output should look similar to the following:
+Deploy the [`cos-lite`](https://charmhub.io/topics/canonical-observability-stack) bundle in a Kubernetes controller following the [MicroK8s guide](https://charmhub.io/topics/canonical-observability-stack/tutorials/install-microk8s). 
```
Question: Do we want to mention `cos-lite` but link to COS instead? There is a dedicated link for COS-lite too: https://charmhub.io/topics/canonical-observability-stack/editions/lite

izmalk commented on docs/how-to/enable-monitoring.md:
````diff
@@ -29,50 +36,51 @@ prometheus/0*    active    idle   10.1.153.190
-juju switch cos
+While still in the COS model, offer COS interfaces to be cross-model related with the machine model that you are using to deploy the etcd charm:
 
+```text
````
Suggestion: We should at least use `shell` or `bash` here and later for commands, or even switch to the `terminal` for consistency with other pages. Or did you want to avoid syntax highlighting here?

a-velasco replied:
I wanted to avoid syntax highlighting here, since the shell and bash highlighting sometimes does confusing things to juju-specific syntax, but the plan is to switch to `terminal`.

izmalk commented on docs/how-to/enable-monitoring.md:
```diff
@@ -29,50 +36,51 @@ prometheus/0*    active    idle   10.1.153.190
-> In the commands above, `k8s` refers to the controller where the COS-lite bundle is deployed, and `cos` refers to the COS model.
+In the commands above, `k8s` refers to the controller where the COS-lite bundle is deployed, and `cos` refers to the COS model.
+
+## Deploy `grafana-agent`
```
Suggestion: Use a bit more specific naming here. My preference is to use the article `the` if I have to use a literal block inside a heading. Non-blocking, but do let me know what you think.

```suggestion
## Deploy the `grafana-agent` charm
```

izmalk commented on docs/how-to/enable-monitoring.md:
```diff
@@ -199,91 +207,91 @@ The metrics exposed by etcd are detailed in the [upstream etcd documentation](ht
 
 They include alerts for:
-* **etcdMembersDown:**
+* ``etcdMembersDown``:
```
Suggestion:  Correct me if I'm wrong, but I think double backticks are not needed.

izmalk commented on docs/how-to/manage-passwords.md:
````diff
@@ -0,0 +1,27 @@
+
+secret:ctbirhuutr9sr8mgrmpg
+```
+Make note of the string following `secret:`.
````
Suggestion: Here and later add empty lines before and after code blocks. That is a Markdown best practice to do so.

izmalk commented on docs/how-to/manage-persistent-storage.md:
```diff
@@ -208,41 +211,27 @@ charmed-etcd/5  data/4      filesystem  etcd-storage  /var/snap/charmed-etcd/com
+For instructions on how to configure credentials to etcd, refer to the guide [How to manage passwords](#manage-passwords). 
 
-Now it is safe to remove your existing charmed etcd application:
+With a password now configured, it is safe to remove your existing charmed etcd application:
```
Suggestion: Empty lines are needed around code blocks.

izmalk commented on docs/how-to/tls/disable-tls.md:
````diff
@@ -0,0 +1,100 @@
+...
+```
+
+## Disable client-to-server encryption in transit and mutual authentication
````
Suggestion: Can we make this and the next heading a bit shorter than this?

```suggestion
## Disable client-to-server encryption in transit
```

a-velasco replied:
Will address in new PR (with some other potential changes to these headings)

izmalk commented on docs/how-to/tls/manage-private-keys.md:
````diff
@@ -1,38 +1,44 @@
-juju add-secret tls-peer-private-key private-key=$(base64 -w0 private-key.key)
+```{terminal}
+:scroll:
+:input: juju add-secret tls-peer-private-key private-key=$(base64 -w0 private-key.key)
````
Praise: I like your usage of the terminal block. I should use more of them too.

izmalk reviewed (changes requested):
Good work! I suggested some non-critical improvements for headings, home page, landing pages, and syntax. One of the common minor issues is missing empty lines around headings or code blocks. Nothing serious.

izmalk commented on docs/index.md:
```diff
@@ -0,0 +1,25 @@
+
+Etcd is a distributed, reliable key-value store for the most critical data of distributed systems. This charmed operator deploys and operates etcd on virtual machines.
+
+Charmed Etd is equipped with several features to securely store and scale complicated data workloads, including TLS encryption, horizontal scaling, password rotation, and easy integration with client applications.
```
Suggestion: A small typo. This should be detectable by the spellchecker.

```suggestion
Charmed Etcd is equipped with several features to securely store and scale complicated data workloads, including TLS encryption, horizontal scaling, password rotation, and easy integration with client applications.
```

a-velasco committed:
Apply review suggestions

skourta reviewed (dismissed):
Thank you @a-velasco

a-velasco committed:
small typos and formatting

reneradoi reviewed (dismissed):
Re-approval

a-velasco committed:
Update docs/how-to/enable-monitoring.md

Co-authored-by: Vladimir Izmalkov <48120135+izmalk@users.noreply.github.com>

a-velasco committed:
Update docs/tutorial/index.md

Co-authored-by: Vladimir Izmalkov <48120135+izmalk@users.noreply.github.com>

a-velasco committed:
Update enable-monitoring.md

skourta committed:
[DPE-6706] - Add Client Relation docs (#62)

### Documentation Enhancements:

*
[`docs/how-to/client-relations.md`](diffhunk://#diff-9d69e9be4a9d29b62682ea8b654a5798f6281012b632109697c32dbccf702220R1-R160):
Added a new guide on integrating etcd with an application, including
instructions for using the `etcd_client` interface and the
`data-integrator` charm. This guide covers configuration steps, code
implementation, and commands for establishing and removing relations.
*
[`docs/how-to/index.md`](diffhunk://#diff-a62277b14aab6742697a72ebe63db2be1ea5ba095495f9c7a425d9fc7b6e8f18R12):
Updated the index to include a link to the new client relations
integration guide.

skourta reviewed (approved):
Re-approval

skourta reviewed (dismissed):
Re-approval

a-velasco committed:
Change main branch to 3.5/edge in configs

a-velasco committed:
Merge branch 'sphinx-integration' of https://github.com/canonical/charmed-etcd-operator into sphinx-integration

Mehdi-Bendriss reviewed (approved):
Thank you Andreia! Great work!

reneradoi reviewed (approved):
Re-approval
